### PR TITLE
#9 Created a new branch Issues/9 

### PR DESCRIPTION
**Pull request by salvadore-jefferson**
*Updated the Instrument constructor to throw NullPointerException

*Added JUnit and Hamcrest dependencies to the POM.xml file

*Created a unit test for Instrument named InstrumentTest, and stored it
in src/test/java.

*Created a test of the hashCode() method to ensure that a specially
constructed Instrument instance returns the expected value.
    
*Created a test of the equals() method to ensure that two separate
instance with the same identifier return true for equals().

*Created a test that two separate instances with different identifiers
return false for equals().

*Created a test for the constructor of Instrument to ensure that a
NullPointerException is thrown if you pass null in the constructor for
the instrument name.

*************************************************
Notes:
All my test seemed to produce the desired results.

One thing I could not get to do exactly what I wanted was to have the
NullPointerException thrown AND still have a clean test. As it is now,
the NullPointerException is thrown, but I get 1 error in my test
results.

### CODE REVIEW

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 1]
```diff
@@ -24,8 +37,38 @@ public String getName() {
...
+      return false;
+    }
+    if (getClass() != otherObject.getClass()) {
+      return false;
+    }
+    Instrument testObject = (Instrument) otherObject;
+    if (instrumentKey != testObject.instrumentKey) {
+      return false;
+    }
+    if (instrumentName != testObject.instrumentName) {
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Is this comparing what you think it is comparing?

Review string comparison, and Java object comparison in general. For primitive values, you can use `!=`, but for actual object instances the `!=` will in fact compare their _identities_, that is, whether or not they are the same instance of an object. Instead you want to use the `equals()` method --- in fact, that's why we're implementing `equals()` for your custom object.

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 2]
```diff
@@ -24,8 +37,38 @@ public String getName() {
...
+      return false;
+    }
+    return hashCode() == testObject.hashCode();
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: The `equals()` method should be _consistent_ with `hashCode()`, but it usually shouldn't _contain_ a call to `hashCode()`. The `equals()` method is to see if the internal properties of the two objects are the same. The `hashCode()` method is to give a number that helps group objects as map keys, for example.

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 3]
```diff
@@ -24,8 +37,38 @@ public String getName() {
...
+    }
+    if (getClass() != otherObject.getClass()) {
+      return false;
+    }
+    Instrument testObject = (Instrument) otherObject;
+    if (instrumentKey != testObject.instrumentKey) {
+      return false;
+    }
+    if (instrumentName != testObject.instrumentName) {
+      return false;
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Did you mean to return `false` here? So the only way this method would return `true` is if the two objects were the same instance?

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 4]
```diff
@@ -24,8 +37,38 @@ public String getName() {
...
+    }
+    return hashCode() == testObject.hashCode();
+
+  }
+
+  @Override
+  public int hashCode() {
+    int result = 17;
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: This is nice and correct and it shows that you've been reading up on how to create hash codes, which is good. It's also good you know how to do this. But nowadays we don't want to remember how to use this algorithm --- just use the hash code utility method in `Objects`. Check that document I gave you.

File: pom.xml [hunk 5]
```diff
@@ -1,23 +1,48 @@
...
+		</dependency>
+		<dependency>
+			<groupId>junit</groupId>
+			<artifactId>junit</artifactId>
+			<version>4.12</version>
+			<scope>test</scope>
+		</dependency>
+		<dependency>
+			<groupId>org.hamcrest</groupId>
+			<artifactId>hamcrest-library</artifactId>
```
Comment (garretwilson) [inline pom.xml]: Double-check the document I gave you. I think there is another Hamcrest dependency you can pull in that only pulls in the necessary stuff, since you're already getting some stuff with JUnit. I may be wrong --- I don't have that document handy.

File: src/test/java/beatbox/InstrumentTest.java [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+import com.github.salvadore_jefferson.beatbox.Instrument;
+
+public class InstrumentTest {
+
+  @Test
+  public void testOverridenHashCode() throws Exception {
+    Instrument testInstrument = new Instrument("Bass Drum", 35);
+    int testHash = testInstrument.hashCode();
+    assertThat(testInstrument.hashCode(), is(testHash));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: I don't understand what you're testing here. You seem to be testing that the `hashCode()` method returns the same value if we call it twice in a row. I don't know how useful that test is.

File: src/test/java/beatbox/InstrumentTest.java [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  @Test
+  public void testTwoInstrumentsAreEqual() throws Exception {
+    Instrument testInstrument = new Instrument("Bass Drum", 35);
+    Instrument testInstrument2 = testInstrument;
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: You are not creating a new `Instrument` instance here. You are merely adding another variable and assigning it _the same instance_ you already created. So basically you have two variables referencing the same instance. This test doesn't help us much --- we already know that if we have two variables referencing the same instance, they will be the same instance. This doesn't test that you implemented `equals()` correctly.

File: src/test/java/beatbox/InstrumentTest.java [hunk 8]
```diff
@@ -0,0 +1,46 @@
...
+    assertThat(testInstrument2, is(sameInstance(testInstrument)));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Here I think you are getting confused about what the `equals()` method is all about. It checks to see that two objects are equivalent, i.e. that they contain the same values (or whatever "equivalent" means for this class). It does _not_ check that the objects are the same instance. In many cases they will _not_ be the same instance.

What you should be doing is creating _another_ instance of `Instrument` and then comparing that it is `equal()` to the first. In other words, `equals()` should  be called at some point. (Hint: The Hamcrest `is()` method will eventually call your object's `equals()` method.)

Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: P.S. When you get this right, the test will fail, exposing your bug in `equals()`.

File: src/test/java/beatbox/InstrumentTest.java [hunk 9]
```diff
@@ -0,0 +1,46 @@
...
+
+  }
+
+  @Test
+  public void testTwoInstrumentsAreNotEqual() throws Exception {
+    Instrument test1 = new Instrument("Bass Drum", 35);
+    Instrument test2 = new Instrument("Closed Hi-Hat", 42);
+    assertThat(test1.hashCode(), is(not(equalTo(test2.hashCode()))));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: You made this test correctly! But I may have asked for the wrong thing. (I don't remember which I asked for.) It turns out that the rules for `hashCode() say that the hash code doesn't _have_ to be different for two unequal objects (although that's good to have). The requirement is that two hash codes have to be the same if the objects are the same. So let's change this test to create two equivalent objects and make sure their hash codes are the same.

File: src/test/java/beatbox/InstrumentTest.java [hunk 10]
```diff
@@ -0,0 +1,46 @@
...
+  }
+
+  @Test
+  public void testInstrumentConstructorForNullPointerException() throws Exception {
+    Instrument test = new Instrument(null, 30);
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Yep! That's what you should be doing. But as you point out, this throws an exception. That's what we want it to do, so somehow we need to tell JUnit that we expect an exception here. I think I told how to do this in a summary point in that document I gave you.

File: src/test/java/beatbox/InstrumentTest.java [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+    assertThat(test.getName(), is(nullValue()));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: We don't want to assert that the name is `null`, because the whole point of our fail-fast precondition is that it will prevent an instrument with a `null` name from ever being created!

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 12]
```diff
@@ -24,8 +37,38 @@ public String getName() {
...
+      return true;
+    }
+    if (otherObject == null) {
+      return false;
+    }
+    if (getClass() != otherObject.getClass()) {
+      return false;
+    }
+    Instrument testObject = (Instrument) otherObject;
+    if (instrumentKey != testObject.instrumentKey) {
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Here we need to think about what it means for two instruments to be "equal". It all depends on how we use the class, and what "equals" means in that context.

Here one question would be, "would there ever be two instruments with different names but with the same number"? I think the answer is no, because the number serves as a sort of unique ID for instruments. So in identifying instruments, the name is not as important as the unique identifier.

For an object such as a "fraction" (with a "numerator" and a "denominator"), there is no "unique ID" as such, so we would want to compare both values. But if our instrument class has a unique ID, our job becomes easier --- just compare they key here and ignore the name! This means that if two instruments have different names but the same key, we still will consider them equal. That will probably be OK, because the key is what is important --- the instrument will play the same regardless of the name as long as it has the same key.

It also makes the `hashKey()` method much easier. If the key is the unique ID, and will be unique for each instrument, we can just use the "key" as the hash key!!

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 13]
```diff
@@ -11,12 +16,20 @@
...
-  public Instrument(String name, int number) {
-    this.instrumentName = name;
-    this.instrumentNumber = number;
+  /**
+   * Instrument constructor for creating MIDI Instruments. Requirements:
+   * 
+   * @param instrumentName
+   * @param instrumentKey
+   */
+  public @Nullable Instrument(@Nonnull final String instrumentName,
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: How can a constructor return `null`? In fact we don't need to tell what a constructor returns. But your annotations for the parameters are nice and spot on!

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 14]
```diff
@@ -11,12 +16,20 @@
...
+      @Nonnull final int instrumentKey) throws NullPointerException {
+    checkNotNull(instrumentName, "You must provide a name for this Instrument");
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: This is fine. But you'll note that `checkNotNull()` returns the instance that you're checking, so you can do the check and the assignment on the same line to save space. Plus it looks nicer.

File: src/test/java/beatbox/InstrumentTest.java [hunk 15]
```diff
@@ -0,0 +1,46 @@
+package beatbox;
+
+import static org.hamcrest.MatcherAssert.assertThat;
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Double-check this. I think there is another `assertThat` that we can use. It may be part of JUnit already. In fact, if you pull in the dependencies correctly, you should be able to just do two `*` static imports to get most everything. See the document I gave you.

File: src/test/java/beatbox/InstrumentTest.java (hunk 11 above)
Comment (salvadore-jefferson) [inline src/test/java/beatbox/InstrumentTest.java]: Thanks for the feedback Garret.
I managed to work on this list and pushed a updated commit to GitHub. For whenever you get a chance to review

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java [hunk 16]
```diff
@@ -24,8 +36,31 @@ public String getName() {
...
+    Instrument testObject = (Instrument) otherObject;
+    if (instrumentKey != testObject.instrumentKey) {
+      return false;
+    }
+    return true;
+  }
+
+  @Override
+  public int hashCode() {
+    return Objects.hash(instrumentKey);
```
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Home work assignment: send me a private email explaining to me, in your own words, what the `hashCode()` method is for. Don't stress over whether you are right or wrong. Just study the Javadocs of the `Object.hashCode()` method, read some documents online, and then tell me what you think it means.

File: src/test/java/beatbox/InstrumentTest.java [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+
+import com.github.salvadore_jefferson.beatbox.Instrument;
+
+public class InstrumentTest {
+
+  @Test
+  public void testOverriddenHashCode() throws Exception {
+    Instrument testInstrument = new Instrument("Bass Drum", 35);
+    Instrument testInstrument2 = new Instrument("Guitar", 35);
+    assertThat(testInstrument, is(equalTo(testInstrument2)));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: This is not a bad test assertion. But it's best to keep the tests as specific as possible to the one thing being tested, so that when it fails we'll automatically know what was wrong. A couple of lines below you correctly test the hash code. This line testing instrument equality needs to go down in your `testTwoInstrumentsAreEqual` test.

File: src/test/java/beatbox/InstrumentTest.java [hunk 18]
```diff
@@ -0,0 +1,45 @@
...
+    assertThat(testInstrument2, is(equalTo(testInstrument)));
+    assertThat(testInstrument.hashCode(), is(equalTo(testInstrument2.hashCode())));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Yes, that's correct! _Tip:_ As a shortcut, you could have used `... is(testInstrument2.hashCode())`. It means the same thing, But it doesn't matter --- whichever way you like better.

File: src/test/java/beatbox/InstrumentTest.java [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+  }
+
+  @Test
+  public void testTwoInstrumentsAreNotEqual() throws Exception {
+    Instrument test1 = new Instrument("Bass Drum", 35);
+    Instrument test2 = new Instrument("Closed Hi-Hat", 42);
+    assertThat(test1.hashCode(), is(not(equalTo(test2.hashCode()))));
+  }
+
+  @Test(expected = NullPointerException.class)
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Nice! You found it. You probably see that the test is passing now.

File: src/test/java/beatbox/InstrumentTest.java [hunk 20]
```diff
@@ -0,0 +1,45 @@
...
+    assertThat(testInstrument, is(equalTo(testInstrument2)));
+    assertThat(testInstrument2, is(equalTo(testInstrument)));
+    assertThat(testInstrument.hashCode(), is(equalTo(testInstrument2.hashCode())));
+  }
+
+  @Test
+  public void testTwoInstrumentsAreEqual() throws Exception {
+    Instrument testInstrument = new Instrument("Bass Drum", 35);
+    Instrument testInstrument2 = new Instrument("Violin", 35);
+    testInstrument2.equals(testInstrument);
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: You are now correctly calling the `equals()` method. But you are not testing anything! This is just a method call --- you aren't doing any assertions. (Note that you actually put the correct call up in `testOverriddenHashCode` --- just bring that line down here.)

File: src/test/java/beatbox/InstrumentTest.java (hunk 9 above)
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Let's take this test out altogether, because as it turns out, two no-equal objects are not required to have different hash codes, so we shouldn't be testing that. Once you understand the hash code concept a bit better, you'll understand why.

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java (hunk 16 above)
Comment (salvadore-jefferson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Welcome back Garret. Again thanks for all your help over your vacation.

Here is my homework :)

The hashCode() method returns an integer representation of the Object's
internal address.

The hash code allows algorithms and data structures, such as hashMap and
hashSet, to put objects into compartments (buckets) for faster access of
the Object. The alternative being an unsorted set of Objects.

Unequal objects do not have to have distinct hashCode values, nor do
objects with the same hashCode have to be equal.

When overriding the equals() method you must also override the hashCode()
method to preserver the hashCode contract.

The hashCode for an Object must be consistent for the duration of an
application run cycle, meaning if the hashCode() method is called multiple
times per application execution then it must return the same integer,
unless the equals comparison has been changed. But this integer can change
for different application executions.

Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: Hi, thanks for the homework, and for the separate email.

The trick I was going to get at is that, since `instrumentKey` (the only value you compare in `equals()`) is an `int` anyway, you can just return the `instrumentKey` value itself instead of getting its `hashCode()`!

File: src/test/java/beatbox/InstrumentTest.java (hunk 15 above)
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: You mentioned elsewhere something about not being able to do wildcard static imports because of the Google style restrictions, and that's OK. I think that Eclipse will give you various options for `assertThat()`. I think the one I always use is `org.junit.Assert.assertThat()`, because `org.junit.Assert` gives me various assertion routines. But `org.junit.Assert.assertThat()` in turn calls `org.hamcrest.MatcherAssert.assertThat()`, so it really doesn't matter. But if you can use wildcards, `import static org.junit.Assert.*` will get you a lot of bang for the buck.

So in short what you have here is fine. I'm just providing extra tips.

File: src/test/java/beatbox/InstrumentTest.java [hunk 21]
```diff
@@ -0,0 +1,36 @@
...
+
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+import org.junit.Test;
+
+import com.github.salvadore_jefferson.beatbox.Instrument;
+
+public class InstrumentTest {
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Where are the Javadocs for this class? At least tell me what this class is for and who wrote it. Remember you can use things like `{@link Instrument}` and `@see Instrument` and `@author`.

File: src/test/java/beatbox/InstrumentTest.java [hunk 22]
```diff
@@ -0,0 +1,36 @@
...
+    Instrument testInstrument2 = new Instrument("Guitar", 35);
+    assertThat(testInstrument.hashCode(), is(equalTo(testInstrument2.hashCode())));
+  }
+
+  @Test
+  public void testTwoInstrumentsAreEqual() throws Exception {
+    Instrument testInstrument = new Instrument("Bass Drum", 35);
+    Instrument testInstrument2 = new Instrument("Guitar", 35);
+    assertThat(testInstrument, is(equalTo(testInstrument2)));
+    assertThat(testInstrument2, is(equalTo(testInstrument)));
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Ah, nice, you're checking equality both ways. I didn't even think of that when I asked for this. Reverse equality is important---good work.

File: src/test/java/beatbox/InstrumentTest.java [hunk 23]
```diff
@@ -0,0 +1,36 @@
...
+
+  }
+
+  @Test(expected = NullPointerException.class)
+  public void testInstrumentConstructorForNullPointerException() throws Exception {
+    Instrument test = new Instrument(null, 30);
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: To prevent a compiler or IDE warning here you can just call `new` without assigning it to anything. Or add a `@SuppressWarnings("unused")` annotation, but that's ugly and unnecessary here.

File: src/test/java/beatbox/InstrumentTest.java [hunk 24]
```diff
@@ -0,0 +1,47 @@
+package beatbox;
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: @salvadore-jefferson , I just noticed this. The `InstrumentTest` class should be in the same package (resulting in the same directory structure) as the class (`Instrument`) it is testing. The only difference is that it is in the `src/test/java` subtree instead of `src/main/java`. Please move it.

File: src/test/java/beatbox/InstrumentTest.java [hunk 25]
```diff
@@ -0,0 +1,47 @@
...
+import static org.hamcrest.MatcherAssert.assertThat;
+import static org.hamcrest.Matchers.equalTo;
+import static org.hamcrest.Matchers.is;
+
+import org.junit.Test;
+
+import com.github.salvadore_jefferson.beatbox.Instrument;
+
+/**
+ * A test class. Used for testing instances of the Instrument class.
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Make it easy for the person reading the comment to get to the `Instrument` class. Add a Javadoc link to it. Tip: Put your cursor at the end of the word `Instrument` and press `Ctrl+Space`.

File: src/test/java/beatbox/InstrumentTest.java [hunk 26]
```diff
@@ -0,0 +1,47 @@
...
+ * 
+ * @version 1.0.0 2015-07-20
+ * @author Salvadore Jefferson
+ */
+
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: All these empty lines between the Javadoc and the class are distracting.

File: src/test/java/beatbox/InstrumentTest.java [hunk 27]
```diff
@@ -0,0 +1,47 @@
...
+
+public class InstrumentTest {
+
+  /** {@link com.github.salvadore_jefferson.beatbox.Instrument#hashCode()}. */
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Great idea to link to the methods being tested. There's nothing technically wrong with including the entire package name, but if the test would have been in the same package then you wouldn't need that part and Eclipse would not have added it.

File: src/test/java/beatbox/InstrumentTest.java [hunk 28]
```diff
@@ -0,0 +1,47 @@
...
+    Instrument testInstrument2 = new Instrument("Guitar", 35);
+    assertThat(testInstrument, is(equalTo(testInstrument2)));
+    assertThat(testInstrument2, is(equalTo(testInstrument)));
+
+  }
+
+  /** {@link com.github.salvadore_jefferson.beatbox.Instrument#Instrument(String str, int num)}. */
+  @Test(expected = NullPointerException.class)
+  public void testInstrumentConstructorForNullPointerException() throws Exception {
+    Instrument test = new Instrument(null, 30);
```
Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Let's get rid of the assignment so we won't have compiler warnings. Just create the new `Instrument` instance.

File: src/test/java/beatbox/InstrumentTest.java (hunk 27 above)
Comment (salvadore-jefferson) [inline src/test/java/beatbox/InstrumentTest.java]: So I added the entire package path (at first anyway) because I thought, if I didn't then you would tell me I should have. That is when I noticed the Test class was in a different package.

I thought adding this class to the test folder was where it should have gone. I still got a lot to learn

Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: What is important is for Javadoc to find the `Instrument` class, because you're linking to it. It is not wrong to include the entire package. I'm just pointing out that because you didn't put the `InstrumentTest` class in the same package as the `Instrument` class, then in this case you _have_ to include the entire package, because they are in different packages and otherwise Javadoc wouldn't find the `Instrument` class!

Once you put the `InstrumentTest` in the same package as `Instrument` (and note that I said "in the same package", not in the same actual directory --- remember that the tests are in a different directory tree), then you have the option of including the full package here or not in the Javadoc. I prefer not to. You may like to. Either way will work, once you move the `InstrumentTest` class to the correct package.

File: src/test/java/beatbox/InstrumentTest.java (hunk 25 above)
Comment (salvadore-jefferson) [inline src/test/java/beatbox/InstrumentTest.java]: I really thought that this would be the case with "@link" (actually jumping to what is linked to, in source code)
And maybe I am not getting it just right, but as it stands, this just shows the documentation of the class or method I am referencing with the @link statement.

Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: I'm not sure what your question is. I was just suggesting that you put `{@link Instrument}` instead of just `Instrument`, to be helpful to the reader. (It also removes the ambiguity of which Instrument you're referring to, because if it were an `Instrument` in another class, you would include the entire package.)

Of course `{@link Instrument}` won't work until you move the test to the correct package.

Comment (salvadore-jefferson) [inline src/test/java/beatbox/InstrumentTest.java]: What I was thinking would happen, say for example, someone noticed an error when testing the 
testOverriddenHashCode() method, they would be able to clink the @link statement and eclipse might jump them over to the class and method that was linked to. For quick access to the method for modifications.
Maybe that is not what is really happening with the @link. And there is a keyboard shortcut similar this already.

Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Yes, you already have `@link` in the method Javadocs. That's good! Here I'm referring to the class Javadoc.

Comment (salvadore-jefferson) [inline src/test/java/beatbox/InstrumentTest.java]: Okay, so I have used your keyboard shortcut to add an @link, in the InstrumentTest class that is linked to the Instrument class. I did this after you instructed me to do so. I want to confirm that is what you are talking about here?

/**
 * A test class. Used for testing instances of the {@link Instrument} class.
 * /

Comment (garretwilson) [inline src/test/java/beatbox/InstrumentTest.java]: Yes! I don't care if you use a keyboard shortcut or type it manually. I just wanted the `@link` there.

File: src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java (hunk 16 above)
Comment (garretwilson) [inline src/main/java/com/github/salvadore_jefferson/beatbox/Instrument.java]: @salvadore-jefferson, did you see this note? Here you can just return `instrumentKey` rather than creating a hash of the instrument key. Let me know if you do not understand why. (It should be in the materials I gave you.)
